model/parameterSheet.cc

```diff
@@ -45,6 +47,137 @@ namespace
       Pango::show();
     }
   };
+}
+
+void ParameterSheet::populateItemvector() {
+  itemvector.clear();	
+  minsky().canvas.selection.recursiveDo(&GroupItems::items,
+                                        [&](Items&, Items::iterator i) {                                 
+                                          itemvector.emplace_back(*i);
+                                          return false;
+                                        });   	
+}	
+
+void ParameterSheet::draw(cairo_t* cairo) const
+{   
+  cairo_rectangle(cairo,0,0,m_width,m_height);
+  cairo_stroke_preserve(cairo);
+  cairo_clip(cairo);
+
+  try
+    {	
+      		
+      //auto& value=ports[0]->getVariableValue();
+      if (!itemvector.empty())
+      //if (item)
+        if (auto v=itemvector[0]->variableCast())
+        //if (auto v=item->variableCast())
+          if (v->type()==VariableType::parameter)
+            {
+	          auto value=v->vValue();
+              Pango pango(cairo);
+              if (value->hypercube().rank()>2)
+                {
+                  pango.setMarkup("Error: rank>2");
+                  cairo_move_to(cairo,0.5*pango.width(),0.5*pango.height());
+                  pango.show();
+                }
+              else
+                {
+                  float x0=0.0, y0=0.0;//+pango.height();
+                  float x=x0, y=y0;
+                  double colWidth=0;
+                  pango.setMarkup("9999");
+                  //float rowHeight=pango.height();
+                  if (value->hypercube().rank()==0)
```

Comment:
This looks like copy/paste from the Sheet widget. Please consider refactoring this out into some common code to keep things DRY.

Reply:
I will, thanks. I just want to get a rudimentary version working for the parameters. My idea is to crop the borders of the rectangles which contain the 2-D hypercube slices to the ink extents, and then to print successive parameter sheets down the parameter tab so that one can scroll to view them. I just need to get ink extents right. For some reason, neither        

               cairo_recording_surface_ink_extents(surface->surface(),0,0,&width,&height);    
nor
               cairo_recording_surface_ink_extents(cairo_get_target (cairo),0,0,&width,&height); 

nor 
                      width = cairo_image_surface_get_width(cairo_get_target(cairo));                                                                 
                      height = cairo_image_surface_get_height(cairo_get_target(cairo));
   
work. they just return 0 for the width and height. Obviously, accessing the surface that has been drawn on is not as straightforward as I thought it would be.

So far, I have only made progress at getting the width as in the attached image. The latest commit reproduces this when I import [this CSV file](https://www.stats.govt.nz/assets/Uploads/Research-and-development-survey/Research-and-development-survey-2019/Download-data/research-and-development-survey-2019-csv.csv
) 

I used cairo_stroke_extents() and cairo_path_extents() for this. Now, if I could only get the correct vertical extents...

![parameter_tab_partial](https://user-images.githubusercontent.com/46541934/84707945-dfd12100-af5f-11ea-87b3-a18feed2c4c2.png)

Reply:
Ink extents should be the go. surface_get_width et al will only return the width you specified when creating the surface, which is likely to be infinite. You may have some luck using the ecolab::cairo::Surface wrapper, which include accessors working out some of this stuff.